### DISCUSSION TITLE
How can I disable HTML rendering?

### DISCUSSION THREAD

**Discussion by 6f6d6172** [getredash/redash#6615] [Q&A]
Rendering HTML from results can sometimes be nice, but more often than not, it's not what I want. When your database contains untrusted input from a web application that may contain HTML, the last thing I want is for Redash to render it automatically -- even if the HTML is sanitized. The sanitization process allows arbitrary images and stylesheets to be loaded, which means selecting content from the database might end up rendering abusive or otherwise offensive content on the screen, or even be used to deliver malicious content. The feature is neat but frankly it's inappropriate to have enabled by default.

Attached is a screenshot of a row that contains CSS that sets every element's background to an animated gif of a cat jumping out of a box. It's clear I put that in there myself, but when running a web application that accepts content from the general public, it's not an entirely unrealistic situation.

<img width="1349" alt="Screenshot 2023-11-22 at 2 35 08 PM" src="https://github.com/getredash/redash/assets/47008591/69c23343-f476-47d7-82ab-209b6b65b52e">

There doesn't appear to be much documentation around HTML rendering, nor does there appear to be a setting to disable this within a single query or as a whole for the org. Is there something I'm missing? I searched issues and discussions and couldn't find anything on the topic either.

**Comment by comur** [comment]
Hello

Does someone have a workaround for that ?

**Comment by gaecoli** [comment]
<img width="410" alt="image" src="https://github.com/getredash/redash/assets/46274612/feaa49d5-bc2d-4016-967e-6428838edabf">

Just disable this button.

**Comment by gaecoli** [comment]
<img width="180" alt="image" src="https://github.com/getredash/redash/assets/46274612/e9bf675e-cc2d-4dc1-b0f3-c4d03433b9e6">

First click Edit Visualization under Table, then click on the column you want to turn off HTML rendering, and finally turn off the button above.

**Reply by gaecoli** [reply]
Should we consider turning off HTML rendering by default? @justinclift

**Reply by comur** [reply]
I think it has to be disabled by default because if we have a style applied via HTML contained inside a column we cannot see edit button or all template is broken...

I was able to fix my cas by editing HTML of the template via inspect and then edit and disable HTML but it's painful

**Reply by gaecoli** [reply]
You can turn it off manually first. We will discuss this solution later.

**Comment by wncm** [comment]
I wish there were an option to disable it by default globally. Several times already it misled our analysts by "fixing" things like `&gt;` and `&quot;`.

**Comment by eradman** [comment]
@ezraodio1 proposed a change to change the default for "Allow HTML content" in https://github.com/getredash/redash/pull/7064

**Comment by justinclift** [comment]
If we change the default, it's not going to break people who upgrade their existing installations is it?

**Reply by eradman** [reply]
It will not change existing installations, because `allowHTML` is saved in the visualization

```
redash=# SELECT jsonb_pretty(options->'columns'->0) FROM visualizations WHERE id=1;
            jsonb_pretty
-------------------------------------
 {                                  +
     "name": "name",                +
     "type": "string",              +
     "fixed": false,                +
     "order": 100000,               +
     "title": "name",               +
     "visible": true,               +
     "allowHTML": true,             +
     "displayAs": "string",         +
     "imageWidth": "",              +
     "allowSearch": false,          +
     "imageHeight": "",             +
     "alignContent": "left",        +
     "booleanValues": [             +
         "false",                   +
         "true"                     +
     ],                             +
     "highlightLinks": false,       +
     "linkUrlTemplate": "{{ @ }}",  +
     "imageUrlTemplate": "{{ @ }}", +
     "linkOpenInNewTab": true,      +
     "linkTextTemplate": "{{ @ }}", +
     "linkTitleTemplate": "{{ @ }}",+
     "imageTitleTemplate": "{{ @ }}"+
 }
(1 row)
```

Changing the default may affect users of the API (`POST /api/visualizations`), but the solution is easy: set this option in the payload instead of relying on the default.

**Reply by justinclift** [reply]
Cool.  Doesn't sound like changing the default won't cause any significant problems then. :smile:

**Comment by eradman** [comment]
We changed the default for "AllowHTML content" in #7064. This only affects new visualizations.
Closing!

**Reply by yoshiokatsuneo** [reply]
@eradman 

> This only affects new visualizations.

It looks not true.
The default "Table" visualization automatically created for new query does not have any options(so options is "{}"), so the existing "Table" uses default settings. 
Now, as the default behavior changed to "disabled", most of existing Table visualization is affected by the changing.
( I just wondered why the output for existing query changed and reached this issue.)

I think this is reasonable changing and not so big deal.  But I just want to make clear the situation.
